Adds an s3 list prefixes operator

jarfgit opened the pull request:
- Adds an operator to return a list of prefixes from an S3 bucket
- Updates `list_prefixes()` unit test to assert on a nested dir with a prefix variable
- Removes duplicate calls to `list_keys()` that were in the `test_list_prefixes()` unit test (likely a copy/paste boo boo?)

There are two suggestion from [this conversation](https://github.com/apache/airflow/pull/8464) that I have not included here:
1. Combine or otherwise simplify `s3_list_keys()` and `s3_list_prefixes()` into one - this makes sense to me but I don't quite know how people tend to use these operators or if there is a valid argument for keeping them separate. 

2. Combining all the s3 operators into one file like [gcs.py](https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/gcs.py) - this also makes sense to me, but it's not consistent with the other AWS operators. Might be worth opening a new issue to refactor them all if we want to go in this direction?

Issue Link: #[8448](https://github.com/apache/airflow/issues/8448) 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

jarfgit committed:
Adds an s3 list prefixes operator

Also:
- Updates list_prefixes() unit tests to assert on a nested dir with a prefix variable
- Removes duplicate calls to list_keys() that were in the list_prefixes() unit test (likely a copy/paste boo boo?)

o-nikolas reviewed (commented):
Thanks for the contribution!

I'm seeing a lot of duplicate code here. As called out previously in the issue, I'd like to see this as at least one operator. The hooks are a different story, which I think is outside the scope of your contribution here.

To me it makes sense to keep the current operator for backwards compatibility and also since it's name is nicely generic (i.e. it's `S3ListOperator`, not `S3ListKeysOperator`). I propose we add a new optional boolean parameter to the existing `S3ListOperator`, perhaps something like `include_subfolders` or `include_common_prefixes`. This should be false by default to maintain the current behaviour, but when true it includes common prefixes in the result payload.

The only issue with this that I see is that if folks want *only* the common prefixes, the new parameter will have to be more complex than just a boolean (your conditions in that case would be `keys`, `common_prefixes`, `all`). **Note**: if we consider that usecase uncommon then another option the users could do is call the operator twice; once with just keys and a second time with keys and subfolders and then do a set difference to get just the subfolders.

jarfgit committed:
Add 'recursive' parameter to s3_list operator

	- Returns common prefixes (i.e. subfolders) in addition to files in the npayload when set to True
	- Avoids breaking change by preserving file-only functionality by defaulting to False

Delete s3_list_prefixes operator and test

Refactor list keys s3  hook to return both files and common prefixes

Refactor calls to list_keys hook, unit tests

Refine param documentation in the S3ListOperator

Add more unit test cases to test_list_keys and test_list_prefixes

Clarify a comment

jarfgit committed:
Revert "Add 'recursive' parameter to s3_list operator"

This reverts commit c995e9a64eb94404948976632d957c2ef0af2181 and returns to the original notion of having a separate list_s3_prefixes operator.

jarfgit committed:
Add more list_prefix hook tests

uranusjr commented on tests/providers/amazon/aws/hooks/test_s3.py:
```diff
@@ -124,11 +124,14 @@ def test_list_prefixes(self, s3_bucket):
         bucket.put_object(Key='dir/b', Body=b'b')
+        bucket.put_object(Key='dir/sub_dir/c', Body=b'c')
 
         assert [] == hook.list_prefixes(s3_bucket, prefix='non-existent/')
+        assert [] == hook.list_prefixes(s3_bucket)
         assert ['dir/'] == hook.list_prefixes(s3_bucket, delimiter='/')
-        assert ['a'] == hook.list_keys(s3_bucket, delimiter='/')
-        assert ['dir/b'] == hook.list_keys(s3_bucket, prefix='dir/')
+        assert [] == hook.list_prefixes(s3_bucket, prefix='dir/')
+        assert ['dir/sub_dir/'] == hook.list_prefixes(s3_bucket, delimiter='/', prefix='dir/')
+        assert [] == hook.list_prefixes(s3_bucket, prefix='dir/sub_dir/')
```
Are these changes related to the operator addition?

jarfgit replied:
The calls to `list_keys` were in the `test_list_prefixes` unit test. I didn't immediately see why these calls were there and assumed it might have been a copy / past error. Am I missing something?

uranusjr commented on airflow/providers/amazon/aws/operators/s3_list_prefixes.py:
```diff
@@ -0,0 +1,97 @@
+            self.bucket,
+            self.prefix,
+            self.delimiter,
+        )
```
```suggestion
        self.log.info(
            'Getting the list of subfolders from bucket: %s in prefix: %s (Delimiter %s)',
            self.bucket,
            self.prefix,
            self.delimiter,
        )
```

Typo? 🙂

jarfgit replied:
Oops :)

jarfgit committed:
Update airflow/providers/amazon/aws/operators/s3_list_prefixes.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

o-nikolas commented on airflow/providers/amazon/aws/operators/s3_list_prefixes.py:
```diff
@@ -0,0 +1,97 @@
+        *,
+        bucket: str,
+        prefix: str = '',
+        delimiter: str = '',
```
Does it really make sense to allow the user to not provide these values?

jarfgit replied:
It doesn't :)

I'll admit this was a lazy copy / paste from S3List, so I updated the `prefix` and `delimiter` there as well.

jarfgit replied:
@o-nikolas github is not letting me re-request a review from you for some reason, so pinging here.

o-nikolas replied:
Hey, sorry I missed this message. Code looks good, and I see that it's merged. Congrats :)

o-nikolas commented on tests/providers/amazon/aws/operators/test_s3_list_prefixes.py:
```diff
@@ -0,0 +1,44 @@
+        mock_hook.return_value.list_prefixes.assert_called_once_with(
+            bucket_name=BUCKET, prefix=PREFIX, delimiter=DELIMITER
+        )
+        assert sorted(subfolders) == sorted(MOCK_SUBFOLDERS)
```
Do we need the sort if there is only one item in `MOCK_SUBFOLDERS`?

jarfgit replied:
We do not, updated in the latest push.

jarfgit committed:
Disallow empty strings for prefix and delimiter in both S3ListPrefixes and S3List operators, remove unnecessary sort() in test, fix typo

kaxil committed:
fixup! Disallow empty strings for prefix and delimiter in both S3ListPrefixes and S3List operators, remove unnecessary sort() in test, fix typo

kaxil commented on airflow/providers/amazon/aws/operators/s3_list.py:
```diff
@@ -72,8 +72,8 @@ def __init__(
-        prefix: str = '',
-        delimiter: str = '',
+        prefix: str,
+        delimiter: str,
```
Is this change required?

jarfgit replied:
Actually no. I copied this code for the S3ListPrefixes operator - where you'd presumably want these to be explicit values, but that doesn't necessarily apply to S3List. Let me revert this and push it up. Good catch :)

kaxil replied:
```suggestion
        prefix: str = '',
        delimiter: str = '',
```

kaxil committed:
Apply suggestions from code review
